src/ngraph/op/fused/matmul.cpp

```diff
@@ -102,3 +103,149 @@ shared_ptr<Node> op::MatMul::clone_with_new_inputs(const OutputVector& new_args)
+    {
+        Shape output_shape;
+        Shape arg0_shape_update = arg0_shape;
+        Shape arg1_shape_update = arg1_shape;
+
+        size_t arg0_rank = arg0_shape.size();
+        size_t arg1_rank = arg1_shape.size();
+
+        if (transpose_a && arg0_rank > 1)
+        {
+            swap(arg0_shape_update[arg0_rank - 2], arg0_shape_update[arg0_rank - 1]);
+        }
+        if (transpose_b && arg1_rank > 1)
+        {
+            swap(arg1_shape_update[arg1_rank - 2], arg1_shape_update[arg1_rank - 1]);
+        }
+
+        if (arg0_rank == 1 && arg1_rank == 1)
+        {
+            NGRAPH_CHECK(arg0_shape_update == arg1_shape_update, "Incompatible arg shapes");
+            output_shape = Shape{};
+        }
+        else if (arg0_rank == 1)
+        {
+            // i.e., arg0 shape {3}, arg1 shape{2, 3, 2}, output shape {2, 2}
+            NGRAPH_CHECK(arg0_shape_update[0] == arg1_shape_update[arg1_rank - 2],
+                         "Incompatible arg shapes");
+            arg1_shape_update.erase(arg1_shape_update.begin() + arg1_rank - 2);
+            output_shape = arg1_shape_update;
+        }
+        else if (arg1_rank == 1)
+        {
+            // i.e., arg0 shape {2, 2, 3}, arg1 shape{3}, output shape {2, 2}
+            NGRAPH_CHECK(arg1_shape_update[0] == arg0_shape_update[arg0_rank - 1],
+                         "Incompatible arg shapes");
+            arg0_shape_update.erase(arg0_shape_update.begin() + arg0_rank - 1);
+            output_shape = arg0_shape_update;
+        }
+        else if (arg0_rank == 2 && arg1_rank == 2)
+        {
+            NGRAPH_CHECK(arg0_shape_update[1] == arg1_shape_update[0], "Incompatible arg shapes");
+            output_shape = Shape{arg0_shape_update[0], arg1_shape_update[1]};
+        }
+        else
+        {
+            NGRAPH_CHECK(arg0_shape_update[arg0_rank - 1] == arg1_shape_update[arg1_rank - 2],
+                         "Incompatible arg shapes");
+
+            const auto& broadcast_shapes = builder::get_numpy_broadcast_shapes(
+                {Shape{begin(arg0_shape_update), next(end(arg0_shape_update), -2)},
+                 Shape{begin(arg1_shape_update), next(end(arg1_shape_update), -2)}});
+
+            output_shape = broadcast_shapes.first;
+            output_shape.insert(output_shape.end(), arg0_shape_update[arg0_rank - 2]);
+            output_shape.insert(output_shape.end(), arg1_shape_update[arg1_rank - 1]);
+        }
+
+        return output_shape;
+    }
+
+    template <element::Type_t ET>
+    bool evaluate(const HostTensorPtr& arg0,
+                  const HostTensorPtr& arg1,
+                  const HostTensorPtr& output,
+                  bool transpose_a,
+                  bool transpose_b)
+    {
+        using T = typename element_type_traits<ET>::value_type;
+
+        Shape arg0_shape = arg0->get_shape();
+        Shape arg1_shape = arg1->get_shape();
+        PartialShape output_pshape = output->get_partial_shape();
```

Comment:
Don't worry about `output_pshape`. Just go directly to the `output_shape` computation in the `else` below.

Reply:
The output shape is checked to see if it is already static, if so, no need to do the computation for output shape.

Reply:
Frameworks don't enforce the static shapes, so you might see inputs that don't correspond to the shape computed during shape inference.